File: cli/src/validate/po.rs
use anyhow::{anyhow, Result};
use regex::Regex;
use std::collections::BTreeMap;

pub fn validate_po(content: &str, source: &str) -> Result<()> {
    validate_po_structure(content)?;

    let entries = parse_po_entries(content);
    let Some(header) = entries.iter().find(|entry| entry.msgid.is_empty()) else {
        return Err(anyhow!(
            r#"po file missing header entry (msgid "" with Content-Type)"#
        ));
    };

    let plural_count = extract_plural_forms_count(&header.msgstr);
    if plural_count > 0 {
        for entry in &entries {
            if !entry.has_plural || entry.msgid.is_empty() {
                continue;
            }

            let max_plural = entry
                .plural_msgstr
                .keys()
                .max()
                .copied()
                .unwrap_or(usize::MAX);
            if max_plural == usize::MAX || max_plural + 1 != plural_count {
                return Err(anyhow!(
                    r#"po plural forms mismatch: header declares nplurals={} but entry for "{}" has {} forms"#,
                    plural_count,
                    truncate(&entry.msgid, 40),
                    if max_plural == usize::MAX {
                        0
                    } else {
                        max_plural + 1
                    }
                ));
            }
        }
    }

    if !source.trim().is_empty() {
        validate_format_strings(source, &entries)?;
    }

    let untranslated = entries
        .iter()
        .filter(|entry| {
            entry.msgid.is_empty() && entry.msgstr.is_empty() && entry.plural_msgstr.is_empty()
        })
        .count();
    if untranslated > 0 {
        return Err(anyhow!("po has {untranslated} untranslated entries"));
    }

    Ok(())
}

fn validate_po_structure(content: &str) -> Result<()> {
    let mut state = "";
    let mut has_msgid = false;
    let mut has_msgstr = false;

    for raw_line in content.lines() {
        let line = raw_line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }

        if line.starts_with("msgid ") {
            if has_msgid && !has_msgstr {
                return Err(anyhow!("po entry missing msgstr"));
            }
            has_msgid = true;
            has_msgstr = false;
            state = "msgid";
            ensure_quoted(line, "po msgid missing quoted string")?;
            continue;
        }

        if line.starts_with("msgid_plural ") {
            if state != "msgid" {
                return Err(anyhow!("po msgid_plural without msgid"));
            }
            ensure_quoted(line, "po msgid_plural missing quoted string")?;
            continue;
        }

        if line.starts_with("msgstr") {
            if !has_msgid {
                return Err(anyhow!("po msgstr without msgid"));
            }
            has_msgstr = true;
            state = "msgstr";
            ensure_quoted(line, "po msgstr missing quoted string")?;
            continue;
        }

        if line.starts_with('"') {
            if state.is_empty() {
                return Err(anyhow!("po stray quoted string"));
            }
            continue;
        }

        return Err(anyhow!("po invalid line: {line}"));
    }

    if has_msgid && !has_msgstr {
        return Err(anyhow!("po entry missing msgstr"));
    }

    Ok(())
}

#[derive(Debug, Clone, Default)]
struct PoEntry {
    msgid: String,
    msgstr: String,
    has_plural: bool,
    plural_msgstr: BTreeMap<usize, String>,
}

fn parse_po_entries(content: &str) -> Vec<PoEntry> {
    let mut entries = Vec::new();
    let mut current = PoEntry::default();
    let mut state = String::new();
    let mut plural_index = None;
    let mut in_entry = false;

    let push_current = |entries: &mut Vec<PoEntry>,
                        current: &mut PoEntry,
                        state: &mut String,
                        plural_index: &mut Option<usize>,
                        in_entry: &mut bool| {
        if !*in_entry {
            return;
        }
        entries.push(current.clone());
        *current = PoEntry::default();
        state.clear();
        *plural_index = None;
        *in_entry = false;
    };

    for raw_line in content.lines() {
        let line = raw_line.trim();
        if line.is_empty() || line.starts_with('#') {
            push_current(
                &mut entries,
                &mut current,
                &mut state,
                &mut plural_index,
                &mut in_entry,
            );
            continue;
        }

        if line.starts_with("msgid ") {
            push_current(
                &mut entries,
                &mut current,
                &mut state,
                &mut plural_index,
                &mut in_entry,
            );
            in_entry = true;
            state = "msgid".into();
            current.msgid = extract_quoted(line);
            continue;
        }

        if line.starts_with("msgid_plural ") {
            current.has_plural = true;
            state = "msgid_plural".into();
            continue;
        }

        if line.starts_with("msgstr[") {
            let closing = line.find(']').unwrap_or(7);
            let index = line[7..closing].parse::<usize>().unwrap_or(0);
            plural_index = Some(index);
            state = "msgstr_plural".into();
            current.plural_msgstr.insert(index, extract_quoted(line));
            continue;
        }

        if line.starts_with("msgstr ") {
            state = "msgstr".into();
            current.msgstr = extract_quoted(line);
            continue;
        }

        if line.starts_with('"') {
            let continuation = extract_quoted_raw(line);
            match state.as_str() {
                "msgid" => current.msgid.push_str(&continuation),
                "msgstr" => current.msgstr.push_str(&continuation),
                "msgstr_plural" => {
                    if let Some(index) = plural_index {
                        current
                            .plural_msgstr
                            .entry(index)
                            .and_modify(|value| value.push_str(&continuation));
                    }
                }
                _ => {}
            }
        }
    }

    push_current(
        &mut entries,
        &mut current,
        &mut state,
        &mut plural_index,
        &mut in_entry,
    );

    entries
}

fn validate_format_strings(source: &str, translated_entries: &[PoEntry]) -> Result<()> {
    let source_entries = parse_po_entries(source);
    let format_regex =
        Regex::new(r"%[sdfiu%]|%\([^)]+\)[sdfiu]|\{[0-9]+\}|\{[a-zA-Z_][a-zA-Z0-9_]*\}").unwrap();

    for source_entry in source_entries {
        if source_entry.msgid.trim().is_empty() {
            continue;
        }

        let Some(translated) = translated_entries
            .iter()
            .find(|entry| entry.msgid == source_entry.msgid)
        else {
            continue;
        };

        if translated.msgstr.trim().is_empty() {
            continue;
        }

        for format_string in format_regex.find_iter(&source_entry.msgstr) {
            let format_string = format_string.as_str();
            if !translated.msgstr.contains(format_string) {
                return Err(anyhow!(
                    r#"po format string "{}" in source msgstr for "{}" missing from translation"#,
                    format_string,
                    truncate(&source_entry.msgid, 40)
                ));
            }
        }
    }

    Ok(())
}

fn ensure_quoted(line: &str, message: &str) -> Result<()> {
    if line.matches('"').count() >= 2 {
        Ok(())
    } else {
        Err(anyhow!(message.to_string()))
    }
}

fn extract_quoted(line: &str) -> String {
    let Some(first) = line.find('"') else {
        return String::new();
    };
    extract_quoted_raw(&line[first..])
}

fn extract_quoted_raw(line: &str) -> String {
    serde_json::from_str::<String>(line).unwrap_or_default()
}

fn extract_plural_forms_count(header: &str) -> usize {
    let regex = Regex::new(r"(?m)^Plural-Forms:\s*nplurals=(\d+);").unwrap();
    regex
        .captures(header)
        .and_then(|captures| captures.get(1))
        .and_then(|match_| match_.as_str().parse::<usize>().ok())
        .unwrap_or(0)
}

fn truncate(input: &str, max_len: usize) -> String {
    if input.chars().count() <= max_len {
        return input.to_string();
    }
    input.chars().take(max_len).collect::<String>() + "..."
}

#[cfg(test)]
mod tests {
    use super::validate_po;

    #[test]
    fn validates_valid_po() {
        let source = r#"msgid ""
msgstr ""
"Content-Type: text/plain; charset=UTF-8\n"
"Plural-Forms: nplurals=2; plural=n != 1;\n"

msgid "Hello"
msgstr "Hello"
"#;

        validate_po(source, source).unwrap();
    }
}

File: app/lib/glossia/translations/validate/po.ex
defmodule Glossia.Translations.Validate.Po do
  @moduledoc """
  Validates Gettext PO output. Ported from the CLI (`cli/src/validate/po.rs`):
  structural checks, a header entry, plural-forms consistency, format-string
  preservation vs the source, and an untranslated-entry check.
  """

  @format_regex ~r/%[sdfiu%]|%\([^)]+\)[sdfiu]|\{[0-9]+\}|\{[a-zA-Z_][a-zA-Z0-9_]*\}/

  def validate_po(content, source) do
    with :ok <- validate_structure(content) do
      entries = parse_entries(content)

      case Enum.find(entries, &(&1.msgid == "")) do
        nil ->
          {:error, ~s|po file missing header entry (msgid "" with Content-Type)|}

        header ->
          with :ok <- validate_plurals(entries, plural_count(header.msgstr)),
               :ok <- validate_format_strings(source, entries) do
            validate_untranslated(entries)
          end
      end
    end
  end

  defp validate_structure(content) do
    content
    |> lines()
    |> Enum.reduce_while({:ok, %{has_msgid: false, has_msgstr: false, state: ""}}, fn raw,
                                                                                      {:ok, s} ->
      line = String.trim(raw)

      cond do
        line == "" or String.starts_with?(line, "#") ->
          {:cont, {:ok, s}}

        String.starts_with?(line, "msgid ") ->
          if s.has_msgid and not s.has_msgstr do
            {:halt, {:error, "po entry missing msgstr"}}
          else
            check_quoted(line, "po msgid missing quoted string", %{
              s
              | has_msgid: true,
                has_msgstr: false,
                state: "msgid"
            })
          end

        String.starts_with?(line, "msgid_plural ") ->
          if s.state != "msgid" do
            {:halt, {:error, "po msgid_plural without msgid"}}
          else
            check_quoted(line, "po msgid_plural missing quoted string", s)
          end

        String.starts_with?(line, "msgstr") ->
          if not s.has_msgid do
            {:halt, {:error, "po msgstr without msgid"}}
          else
            check_quoted(line, "po msgstr missing quoted string", %{
              s
              | has_msgstr: true,
                state: "msgstr"
            })
          end

        String.starts_with?(line, "\"") ->
          if s.state == "" do
            {:halt, {:error, "po stray quoted string"}}
          else
            {:cont, {:ok, s}}
          end

        true ->
          {:halt, {:error, "po invalid line: #{line}"}}
      end
    end)
    |> case do
      {:ok, %{has_msgid: true, has_msgstr: false}} -> {:error, "po entry missing msgstr"}
      {:ok, _state} -> :ok
      {:error, _} = error -> error
    end
  end

  defp check_quoted(line, message, state) do
    if line |> String.graphemes() |> Enum.count(&(&1 == "\"")) >= 2 do
      {:cont, {:ok, state}}
    else
      {:halt, {:error, message}}
    end
  end

  defp validate_plurals(_entries, 0), do: :ok

  defp validate_plurals(entries, plural_count) do
    Enum.reduce_while(entries, :ok, fn entry, :ok ->
      if not entry.has_plural or entry.msgid == "" do
        {:cont, :ok}
      else
        forms =
          if map_size(entry.plural_msgstr) == 0,
            do: 0,
            else: Enum.max(Map.keys(entry.plural_msgstr)) + 1

        if forms == plural_count do
          {:cont, :ok}
        else
          {:halt,
           {:error,
            ~s|po plural forms mismatch: header declares nplurals=#{plural_count} but entry for "#{truncate(entry.msgid, 40)}" has #{forms} forms|}}
        end
      end
    end)
  end

  defp validate_format_strings(source, entries) do
    if String.trim(source) == "" do
      :ok
    else
      source
      |> parse_entries()
      |> Enum.reject(&(String.trim(&1.msgid) == ""))
      |> Enum.reduce_while(:ok, fn source_entry, :ok ->
        check_entry_format_strings(source_entry, entries)
      end)
    end
  end

  defp check_entry_format_strings(source_entry, entries) do
    with translated when not is_nil(translated) <-
           Enum.find(entries, &(&1.msgid == source_entry.msgid)),
         false <- String.trim(translated.msgstr) == "" do
      missing =
        @format_regex
        |> Regex.scan(source_entry.msgstr)
        |> Enum.map(&hd/1)
        |> Enum.find(&(not String.contains?(translated.msgstr, &1)))

      if missing do
        {:halt,
         {:error,
          ~s|po format string "#{missing}" in source msgstr for "#{truncate(source_entry.msgid, 40)}" missing from translation|}}
      else
        {:cont, :ok}
      end
    else
      _ -> {:cont, :ok}
    end
  end

  defp validate_untranslated(entries) do
    count =
      Enum.count(entries, fn e ->
        e.msgid == "" and e.msgstr == "" and map_size(e.plural_msgstr) == 0
      end)

    if count > 0, do: {:error, "po has #{count} untranslated entries"}, else: :ok
  end

  # ── parsing ────────────────────────────────────────────────────────────────

  defp parse_entries(content) do
    content
    |> lines()
    |> Enum.reduce(
      %{entries: [], current: new_entry(), state: "", plural_index: nil, in_entry: false},
      &parse_line/2
    )
    |> push_current()
    |> Map.fetch!(:entries)
    |> Enum.reverse()
  end

  defp parse_line(raw, acc) do
    line = String.trim(raw)

    cond do
      line == "" or String.starts_with?(line, "#") ->
        push_current(acc)

      String.starts_with?(line, "msgid ") ->
        acc = push_current(acc)

        %{
          acc
          | in_entry: true,
            state: "msgid",
            current: %{new_entry() | msgid: extract_quoted(line)}
        }

      String.starts_with?(line, "msgid_plural ") ->
        %{acc | state: "msgid_plural", current: %{acc.current | has_plural: true}}

      String.starts_with?(line, "msgstr[") ->
        index = plural_index(line)

        %{
          acc
          | state: "msgstr_plural",
            plural_index: index,
            current: %{
              acc.current
              | plural_msgstr: Map.put(acc.current.plural_msgstr, index, extract_quoted(line))
            }
        }

      String.starts_with?(line, "msgstr ") ->
        %{acc | state: "msgstr", current: %{acc.current | msgstr: extract_quoted(line)}}

      String.starts_with?(line, "\"") ->
        append_continuation(acc, extract_quoted_raw(line))

      true ->
        acc
    end
  end

  defp append_continuation(%{state: "msgid"} = acc, text),
    do: %{acc | current: %{acc.current | msgid: acc.current.msgid <> text}}

  defp append_continuation(%{state: "msgstr"} = acc, text),
    do: %{acc | current: %{acc.current | msgstr: acc.current.msgstr <> text}}

  defp append_continuation(%{state: "msgstr_plural", plural_index: index} = acc, text)
       when not is_nil(index) do
    case Map.fetch(acc.current.plural_msgstr, index) do
      {:ok, existing} ->
        %{
          acc
          | current: %{
              acc.current
              | plural_msgstr: Map.put(acc.current.plural_msgstr, index, existing <> text)
            }
        }

      :error ->
        acc
    end
  end

  defp append_continuation(acc, _text), do: acc

  defp push_current(%{in_entry: false} = acc), do: acc

  defp push_current(acc) do
    %{
      acc
      | entries: [acc.current | acc.entries],
        current: new_entry(),
        state: "",
        plural_index: nil,
        in_entry: false
    }
  end

  defp new_entry, do: %{msgid: "", msgstr: "", has_plural: false, plural_msgstr: %{}}

  defp plural_index(line) do
    with [_, inside] <- Regex.run(~r/^msgstr\[([^\]]*)\]/, line),
         {index, ""} <- Integer.parse(inside) do
      index
    else
      _ -> 0
    end
  end

  defp plural_count(header) do
    case Regex.run(~r/(?m)^Plural-Forms:\s*nplurals=(\d+);/, header) do
      [_, count] -> String.to_integer(count)
      _ -> 0
    end
  end

  defp extract_quoted(line) do
    case :binary.match(line, "\"") do
      {start, _len} -> extract_quoted_raw(binary_part(line, start, byte_size(line) - start))
      :nomatch -> ""
    end
  end

  defp extract_quoted_raw(line) do
    case Jason.decode(line) do
      {:ok, value} when is_binary(value) -> value
      _ -> ""
    end
  end

  defp truncate(input, max_len) do
    if String.length(input) <= max_len do
      input
    else
      String.slice(input, 0, max_len) <> "..."
    end
  end

  defp lines(content) do
    content |> String.split("\n") |> Enum.map(&String.trim_trailing(&1, "\r"))
  end
end
